pennylane_lightning/src/Bindings.cpp

```diff
@@ -59,17 +60,67 @@ template <class fp_t> class StateVecBinder : public StateVector<fp_t> {
                const vector<vector<fp_t>> &params) {
         this->applyOperations(ops, wires, inverse, params);
     }
+    void apply(const vector<string> &ops, const vector<vector<size_t>> &wires,
+               const vector<bool> &inverse) {
+        this->applyOperations(ops, wires, inverse);
+    }
+    /**
+     * @brief Directly apply a given matrix to the specified wires. Matrix data
+     * in 1D row-major format.
+     *
+     * @param matrix
+     * @param wires
+     */
+    void applyMatrixWires(const std::vector<std::complex<fp_t>> &matrix,
+                          const vector<size_t> &wires, bool inverse = false) {
+        this->applyOperation(matrix, wires, inverse);
+    }
+    /**
+     * @brief Directly apply a given matrix to the specified wires. Data in 1/2D
+     * numpy complex array format.
+     *
+     * @param matrix
+     * @param wires
+     * @param inverse
+     */
+    void applyMatrixWires(const py::array_t<complex<fp_t>> &matrix,
+                          const vector<size_t> &wires, bool inverse = false) {
```

Comment:
Yea, that is the plan. The second approach (the `py::array_t` version) works with the way we have everything currently implemented, and allows for as easy as possible a conversion from the numpy array to the C++ layer by casting the pointer to the data buffer, an following through to the underlying implementation accepting pointers over references. The first approach though will be easier to maintain, and adapt. We have both now as both will allow development and comparison to define our needs as the CPP implementations are fleshed out more.

For the moment I think this is the best way for us to handle this, but long-term, I think we may opt to have a seamless CPP backed matrix class that we pass instead. This would alleviate the need for the multiple overloads, but may make interop with other python interfaces a little trickier.

I think this will likely evolve in time once we eventually move away from the fully Python defined implementations.